README
=Gbarcode

This project is a C extension that wraps the GNU Barcode project using SWIG. 
The Gbarcode version coincides with the GNU Barcode version, the source of which is distributed 
along with the gem, as barcode is not usually installed as a dynamically loaded library on 
most systems. You should not need SWIG to be installed to use this gem, only if you wish to muck around with the 
SWIG interface file. See README_SWIG.txt for more information.

This library is distributed under the GPL, see LICENSE.txt.

For recent changes, see CHANGELOG.txt

==USAGE

The Gbarcode class, being a straight wrap of the GNU Barcode C library, is not very Ruby centric. 
The example of the usage will illustrate this point.

	require 'rubygems'
	require 'gbarcode'

	# include the module so I type less ;-)
	include Gbarcode
	
	# There are three stages to creating a barcode
	# 1) allocating the space with a string
	# 2) encoding the string as some barcode
	# 3) printing out the barcode to Postscript
	
	# allocate the barcode for text "TEST1234"
	bc = barcode_create("TEST1234")

	# encode the barcode using code 39,since code 39 does not use a checksum, we can pass in this additional flag.
	barcode_encode(bc, BARCODE_NO_CHECKSUM | BARCODE_39)

	#print out the barcode to a Postscript file, using the default width and height and page placement
	barcode_print(bc, File.new("testout.ps", "w"), 0,0,0,0,0)

	#print out the barcode to a Postscript EPS file, using the default width and height and page placement
	barcode_print(bc, File.new("testout.eps", "w"), 0,0,0,0,BARCODE_OUT_EPS)

	# you can even read the output to a stream like so
	rd, wr = IO.pipe
	barcode_print(bc, wr, 0,0,0,0,BARCODE_OUT_EPS)
	wr.close() # must close this to use the read pipe
	myEpsBarcodeAsString = rd.readlines().join("\n")
	rd.close()	# it is good practice to also close this pipe
	
It is good practice to close any IO pipes, or else you may get over-run by zombie processes when
using with long-running applications (e.g. Rails)


==BARCODE FORMATS

Gbarcode and Rbarcode, being a wrap of GNU Barcode, supports those formats that GNU barcode supports. From the GNU barcode documentation:

The encodings flags are supplied as constants from Gbarcode. Notice the use of the bitwise OR operation on the flag in the USAGE section.
Here are a  listing of the flags, taken from the GNU barcode docs:
	BARCODE_EAN
	BARCODE_UPC
	BARCODE_ISBN
	BARCODE_128B
	BARCODE_128C
	BARCODE_128
	BARCODE_128RAW
	BARCODE_39
	BARCODE_I25
	BARCODE_CBR
	BARCODE_MSI
	BARCODE_PLS
	BARCODE_93
	     The currently supported encoding types: EAN (13 digits, 8 digits,
	     13 + 2 add-on and 13 + 5 add-on), UPC (UPC-A, UPC-E, UPC-A with 2
	     or 5 digit add-on), ISBN (with or without the 5-digit add-on),
	     CODE128-B (the whole set of printable ASCII characters), CODE128-C
	     (two digits encoded by each barcode symbol), CODE128 (all ASCII
	     values), a "raw-input" pseudo-code that generates CODE128 output,
	     CODE39 (alphanumeric), "interleaved 2 of 5" (numeric), Codabar
	     (numeric plus a few symbols), MSI (numeric) and Plessey (hex
	     digits).  *Note Supported Encodings::.

	BARCODE_ANY
	     This special encoding type (represented by a value of zero, so it
	     will be the default) tells the encoding procedure to look for the
	     first encoding type that can deal with a textual string.
	     Therefore, a 11-digit code will be printed as UPC (as well as
	     6-digit, 11+2 and 11+5), a 12-digit (or 7-digit, or 12+2 or 12+5)
	     as EAN13, an ISBN code (with or without hyphens, with or without
	     add-5) will be encoded in its EAN13 representation, an even number
	     of digits is encoded using CODE128C and a generic string is
	     encoded using CODE128B. Since code-39 offers a much larger
	     representation for the same text string, code128-b is preferred
	     over code39 for alphanumeric strings.

	BARCODE_NO_CHECKSUM
	     Instructs the engine not to add the checksum character to the
	     output. Not all the encoding types can drop the checksum; those
	     where the checksum is mandatory (like EAN and UPC) just ignore the
	     flag.


Here are some flags that can be given to the barcode_print method:

	`BARCODE_OUT_PS'
	`BARCODE_OUT_EPS'
	`BARCODE_OUT_PCL'
	`BARCODE_OUT_PCL_III'
	     The currently supported encoding types: full-page postscript and
	     encapsulated postscript; PCL (print command language, for HP
	     printers) and PCL-III (same as PCL, but uses a font not available
	     on older printers).

	`BARCODE_NO_ASCII'
	     Instructs the engine not to print the ascii string on output. By
	     default the bar code is accompanied with an ascii version of the
	     text it encodes.
	
	`BARCODE_OUT_NOHEADERS'
	     The flag instructs the printing engine not to print the header and
	     footer part of the file. This makes sense for the postscript
	     engine but might not make sense for other engines; such other
	     engines will silently ignore the flag just like the PCL back-end
	     does
	
Multiple flags are given by bitwise OR-ing them together (e.g. "BARCODE_NO_CHECKSUM | BARCODE_39" ) For more details on GNU barcode itself, see the GNU Barcode site at http://www.gnu.org/software/barcode/barcode.html .

==FUNCTION REFERENCE

Since RDoc for some reason was not picking up my comments from the C code, here is a reference for the functions exported by the library:


Gbarcode::BarcodeItem = Gbarcode.barcode_create( _text_ )
	Class method for creating a BarcodeItem for the given string
		[text] The barcode text to encode
		+returns+ Gbarcode::BarcodeItem

int = Gbarcode.barcode_delete( Gbarcode::BarcodeItem )
	Class method for destroying the BarcodeItem. Return 0 if successful

int = Gbarcode.barcode_encode( Gbarcode::BarcodeItem, <ENCODING FLAGS> )
	Encodes the BarcodeItem to one of the flags above. Returns 0 if successful.
	
int = Gbarcode.barcode_print( Gbarcode::BarcodeItem, File, <OUTPUT FLAGS>)'
	     Print the bar code described by bc to the specified file.  Valid
	     flags are the output type, BARCODE_NO_ASCII and
	     BARCODE_OUT_NOHEADERS, other flags are ignored. If any of these
	     flags is zero, it will be inherited from bc->flags which therefore
	     takes precedence. The function returns 0 on success and -1 in case
	     of error (with bc->error set accordingly). In case of success, the
	     bar code is printed to the specified file, which won't be closed
	     after use.

int = Gbarcode.barcode_position( Gbarcode::BarcodeItem, int width, int height, int xoff, int yoff, double scalef);'
	     The function is a shortcut to assign values to the BarcodeItem object. Sets the dimensions, page offsets and scaling factors all in one go. See below for more complete documentation of BarcodeItem.

int = Gbarcode.barcode_encode_and_print(char *text, FILE *f, int wid, int hei, int xoff, int yoff, int flags);'
	     The function deals with the whole life of the barcode object by
	     calling the other functions; it uses all the specified flags.

Gbarcode::BarcodeItem.ascii
    Returns the ASCII text that is being encoded

Gbarcode::BarcodeItem.ascii=
	Sets the ASCII text to be encoded. Please note that if the barcode was already encoded, you will
	need to re-encode it once this method has been called, or else you will be printing out the wrong barcode!

Gbarcode::BarcodeItem.encoding
	Gets the encoding scheme as a human readable string

Gbarcode::BarcodeItem.partial
	Gets a representation of the barcode as a sequence of integers representing the width of a span.
	The integers alternate between specifying white space and a black bar, starting with white space.

Gbarcode::BarcodeItem.textinfo
	Gets the Postscript representation of the ASCII text that will be printed to the barcode file.

Gbarcode::BarcodeItem.width=
Gbarcode::BarcodeItem.width
	Set/get the width in points, which are 1/72  of an inch, or about 1 pixel on most monitors

Gbarcode::BarcodeItem.width=
Gbarcode::BarcodeItem.width
	Set/get the height in points

Gbarcode::BarcodeItem.xoff=
Gbarcode::BarcodeItem.xoff
	Set/get the page X coordinate offset 

Gbarcode::BarcodeItem.yoff=
Gbarcode::BarcodeItem.yoff
	Set/get the page Y coordinate offset

Gbarcode::BarcodeItem.margin=
Gbarcode::BarcodeItem.margin
	Set/get the margin around the barcode in points

Gbarcode::BarcodeItem.scalef=
Gbarcode::BarcodeItem.scalef
	Set/get the scaling factor of the barcode



gbarcode-0.98.16-mswin32/CHANGELOG
4/3/2007

* Initial SWIG wrap of GNU Barcode to Ruby. Not very Ruby-ish interface, but what do you want for free ;-) 


test/test_helper.rb
require 'test/unit'
require 'gbarcode'

test/gbarcode_test.rb
require File.dirname(__FILE__) + '/test_helper.rb'

class GbarcodeTest < Test::Unit::TestCase

  def setup 
    @@BC_TEXT = "Gbarcode"
    @@BC = Gbarcode::Barcode.new(@@BC_TEXT, "CODE128B")
  end

  def test_barcode_create
    assert(@@BC != nil, "BC not created")
  end

  def test_ascii
    assert_equal(@@BC.ascii, "Gbarcode")
  end

  def test_to_eps 
    b = Gbarcode::Barcode.new("Gbarcode", "code 128b").to_eps
    f = File.open(File.dirname(__FILE__) + "/assets/gb-code128b.eps").readlines.join("\n")
    assert_equal(b,f)
  end
end


gbarcode-0.98.16-mswin32/test/gbarcode_test.rb
require File.dirname(__FILE__) + '/test_helper.rb'

class GbarcodeTest < Test::Unit::TestCase

  def setup 
    @@BC_TEXT = "Gbarcode"
    @@BC = Gbarcode.barcode_create(@@BC_TEXT)
    Gbarcode.barcode_encode(@@BC, Gbarcode::BARCODE_128B)
  end

  def teardown
    Gbarcode.barcode_delete(@@BC)
  end
  
  def test_barcode_create
    assert(@@BC != nil, "BC not created")
  end
  
  def test_barcode_delete
    r = Gbarcode.barcode_delete(Gbarcode.barcode_create(@@BC_TEXT))
    assert(r == 0, "barcode_delete failed")
  end
  
  def test_ascii
    assert_equal(@@BC.ascii, "Gbarcode")
  end
  
  def test_barcode_encode
    b = Gbarcode.barcode_create("1234")
    r = Gbarcode.barcode_encode(b, Gbarcode::BARCODE_39)
    assert(r == 0, "encoding unsuccessful")
  end
  
  def test_encoding
    assert_equal(@@BC.encoding, "code 128-B")
  end
  
  def test_barcode_print
    r,w = File.pipe
    Gbarcode.barcode_print(@@BC,w,0)
    w.close()
    b = r.readlines().join("\n")
    r.close()
    f = File.open(File.dirname(__FILE__) + "/ assets/gb-code128b.eps").readlines.join("\n")
    assert_equal(b,f)
  end
  
end


gbarcode-0.98.16-mswin32/ext/svg.c
/*
 * svg.c -- print out the barcode SVG-style
 *
 * Copyright (c) 2001 David J. Humphreys (humphrd@tcd.ie)
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 *
 * http://www.onefour.net/barcode/
 * 
 * Made with a mac! (Tested under os X, 10.1)
 * 
 * I wanted to output barcodes as SVG for some art work.
 * This will output everything as an SVG file. This file can
 * then be inserted into another SVG file.
 *
 * I don't want to get into the specifics of SVG files here. I
 * suppose there will be a few specifics on my website (above) along
 * with a few examples.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <errno.h>

#include "barcode.h"

/* If there is any need to print out a comment,
 * then this is the method */
int svg_comment(char *Message, FILE *f){
  fprintf(f, "<!-- %s -->\n",Message);
  return 0;
}

/* Prints the title (bc->ascii) and the 
 * desc(ription) */
int svg_desc(struct Barcode_Item *bc, FILE *f){
  fprintf(f, "<title>%s</title>\n",bc->ascii);
  fprintf(f, "<desc>\n");
    fprintf(f, "\t%s:\t%s\n",bc->encoding,bc->ascii);
    fprintf(f, "\tGNU-barcode:\t%s\n",BARCODE_VERSION);
    fprintf(f, "\tbarcode:\thttp://www.gnu.org/\n");
    fprintf(f, "\tsvg-code:\thttp://www.onefour.net/barcode\n");
  fprintf(f, "</desc>\n");
  return 0;
}

/* print out the start of the svg file,
 * with some xml voodoo.   There are a
 * lot of \"s in this code.
 * It bounds the svg file by
 * bc->width or  104 and
 * bc->height or 100
 * The title of the file is bc->ascii 
 * NOTE: margins and scale are ignored
 *       the user can simply insert the 
 *       svg file into another and then
 *       specify scale and margin.*/
int svg_start( struct Barcode_Item *bc, FILE *f){
  if(bc->height == 0) bc->height = 100;
  if(bc->width == 0)  bc->width = 104;
  fprintf(f, "<?xml version=\"1.0\" standalone=\"no\"?>\n");
  fprintf(f, "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.0//EN\" \"http://www.w3.org/TR/2001/REC-SVG-20010904/DTD/svg10.dtd\">\n");
  fprintf(f, "<svg width=\"%i\" height=\"%i\">\n",bc->width ,bc->height);
  svg_desc(bc,f);
  return 0;
}

/* just close the svg file */
int svg_end(FILE *f){
  fprintf(f, "</svg>\n");
  return 0;
}

/* prints out the code for the bars.
 * the bars are drawn with rect(angles)
 * rather than lines.   The mathematics
 * for placing a line then setting the 
 * stroke width gets messy. */
int svg_bars(struct Barcode_Item *bc, FILE *f){
  int 
    i = 0,			/* Loop counter */
    height = 70,		/* Each box is 70 pixels in hieght */
    xpos = 0;			/* Where the current box is drawn */
  int current;			/* The current character in partial */
  int is_bar = 0;

  fprintf(f, "<g fill=\"black\">\n");

  for(i = 0; i < strlen(bc->partial); i++){
    current = (int)bc->partial[i] - 48;
    if(current > 9){
        height = 75;		/* Guide bar */
        current = (int)bc->partial[++i] - 48;
        i++;			/* Skip the following 'a' */
    }
    else
        height = 70;
    if(is_bar == 1){
        fprintf(f, "\t<rect x=\"%i\" y=\"0\" ",xpos); /* all bars start at y=0 */
        fprintf(f, "height=\"%i\" width=\"%i\" stroke-width=\"0\"/>\n",height,current);
        is_bar = 0;
    }
    else
        is_bar = 1;
    xpos += current;
  }
  fprintf(f, "</g>\n");
  return 0;
}

/* positions the characters that will
 * be printed.   This assumes they will
 * be under the barcode.*/
int svg_text(struct Barcode_Item *bc, FILE *f){
  unsigned 
    i = 0,
    j = 0,
    infosz = strlen(bc->textinfo),
    xpos = 0,
    prev_x = 0,
    correction = 0;		/* This correction seems to be needed to align text properly */
  double dub1, dub2;
  char printer;
  char *temp_info = malloc(infosz);
  if(!temp_info)
    return -1;
  fprintf(f, "<g font-family=\"Helvitica\" font-size=\"12\">\n");
  
  while(i < infosz){
    for(j = 0; bc->textinfo[i + j + 1] != ' ' &&
	  bc->textinfo[i + j + 1] != '\0';j++);	/* Empty loop, just increment j */

    j ++;
    strncpy(temp_info, (bc->textinfo + i),j);
    sscanf(temp_info, "%lf:%lf:%c", &dub1, &dub2, &printer);
    i += j;

    xpos = (int)dub1;
    if((xpos - prev_x) >= 10)
      correction += 2;

    prev_x = xpos;
    fprintf(f, "\t<text x=\"%i\" y=\"%i\">%c</text>\n",(xpos - correction),80,printer);
  }
  fprintf(f, "</g>\n");
  free(temp_info);
  return 0;
}

/* This is the function to call.   It calls the 
 * functions above in order.   Thers is no real
 * error detection in this code. */
int Barcode_svg_print(struct Barcode_Item *bc, FILE *f){
  svg_start(bc,f);
  svg_bars(bc,f);
  svg_text(bc,f);		/* Should test for text output bit in flags */
  svg_end(f);
  return 0;
}


lib/gbarcode/barcode.rb
require 'rmagick'

module Gbarcode
  include Cgbc
  protected :barcode_create, :barcode_delete, :barcode_encode, :barcode_position, :barcode_print,
  :barcode_encode_and_print,  :barcode_version, :barcode_svg_print, :initialize

  @@VERSION = BARCODE_VERSION

  # Exception thrown to signify encoding errors. Maybe a bit too stringent a method, but I didn't particularly feel like learning how 
  # to catching exceptions from C using SWIG! 
  class EncodingException < Exception
  end

  ENCODINGS = 
  {"CODE 39" => BARCODE_NO_CHECKSUM | BARCODE_39,
    "CODE39" => BARCODE_NO_CHECKSUM | BARCODE_39,
    "CODE128" => BARCODE_128,
    "CODE 128" => BARCODE_128,
    "I25"=> BARCODE_I25,
    "INTERLEAVED 2 OF 5"=> BARCODE_I25,
    "INTRLV25"=> BARCODE_I25,
    "INTRLV 25" => BARCODE_I25,
    "CODABAR"=> BARCODE_CBR,
    "CBR" => BARCODE_CBR,
    "MSI" => BARCODE_MSI,
    "PLESSEY" => BARCODE_PLS,
    "PLS" => BARCODE_PLS,
    "EAN" => BARCODE_EAN,
    "UPC" => BARCODE_UPC,
    "ISBN" => BARCODE_ISBN,
    "CODE93" => BARCODE_93,
    "CODE 93" => BARCODE_93,
    "CODE 128B" => BARCODE_128B,
    "CODE128B" => BARCODE_128B,
    "CODE 128C" => BARCODE_128C,
    "CODE128C"=> BARCODE_128C
  }

  # The Main barcoding class, instatiate this with a text string and optional encoding scheme. See link:README.txt for more information
  class Barcode
    include Gbarcode

    # A class method for encoding a string into a barcode and writing the result out to the given filename. 
    # Default encoding = Code 39. The format of the output is taken from the filename's extension. Please note that 
    # RMagick does not do a good job of converting EPS to SVG images. If you want SVG images, use the instance method
    # to_svg instead.
    def self.encode_and_print str, filename, wdth = 0, hght=0, xoff=0, yoff=0, enc="CODE39"
      r,w = IO.pipe
      Rbarcode.barcode_print(str,w, wdth,hght,xoff,yoff, enc_to_const(enc) | BARCODE_OUT_EPS  )
      w.close
      img = Magick::Image.read(r)
      r.close
      img.write(filename)
    end

    def initialize str="", enc=nil
      @bc = barcode_create(str)
      encode(enc)
    end

    #:nodoc:
    def enc_to_const enc
      if ENCODINGS.has_key? enc.upcase
        return ENCODINGS[enc.upcase]
      else
        return  ENCODINGS["CODE39"]
      end
    end
    #:nodoc:
    def create_image
      unless( @img) then
        r,w = IO.pipe
        barcode_print(@bc,w, BARCODE_OUT_EPS)
        w.close
        @img = Magick::Image.read(r)[0]
        r.close
      end
      return self
    end

    public

    # Encodes the barcode into one of the supported schemes. Default encoding is Code 39. See <<README>> for supported types
    def encode enc
      enc = "CODE39" unless enc
      enc_method  = enc_to_const(enc)
      rs = barcode_encode(@bc, enc_method)
      if (rs == 0 ) then 
        create_image()
      else
        raise EncodingException.new("#{self.class}: Text \"#{@bc.ascii}\" could not be encoded as \"#{enc}\"")
      end
      return self
    end

    # Returns the ascii text being encoded
    def ascii
      @bc.ascii
    end

    # sets the ascii text being encoded. If this is called
    def ascii= str
      @bc.ascii = str
      encode(@bc.encoding)
    end

    def width
      @bc.width
    end

    # Sets the width of the barcode image (in points 1/72 of an inch)
    def width= w
      @bc.width = w
    end

    # Get the height of the barcode image, if set (in points: 1/72 of an inch) 
    def height
      @bc.height
    end

    # Sets the height of the barcode image (in points 1/72 of an inch)
    def height= h
      @bc.height = h
    end

    # Retrieve a text encoding of the barcode as integers. Each integer represents a spacing, where they 
    # alternate between space and black fills (bars), starting with a leading space. The int value represents 
    # the relative width of a bar/space. 
    # Example : 03132112 = no leading space + <tt>XXX_XXX__X_XX</tt> (where "_" = space and "X" = bar)
    def partial
      @bc.partial
    end

    def to_s
      return "#{self.class}: (#{@bc.encoding}) #{@bc.ascii}"
    end

    # Creates an SVG image for the barcode. If no filename is given, output will be returned as a string
    def to_svg fname = nil
      if fname 
        w = File.open(fname,'w')
        barcode_svg_print(@bc,w)
        w.close()
        return self
      else
        r,w = IO.pipe
        barcode_svg_print(@bc,w)
        w.close()
        svg = r.readlines().join("\n")
        r.close
        return svg
      end
    end

    # Write the image out to a file of the given name. Format of the output is taken from the extension.
    def print_to_file fname
      @img.write(fname)
    end

    #:nodoc:all
    def to_imgformat fname=nil, frmt=nil
      if fname 
        @img.write(fname)
        return self
      elsif frmt 
        @img.format = frmt
        return @img.to_blob
      end
    end

    # Write out the barcode to an EPS image. If no filename is given, output will be returned as a string.
    def to_eps fname = nil
      if fname
        w = File.open(fname,'w')
        barcode_print(@bc,w, BARCODE_OUT_EPS)
        w.close()
        return self
      else
        r,w = IO.pipe
        barcode_print(@bc,w, BARCODE_OUT_EPS)
        w.close
        tmp =  r.readlines().join("\n")
        r.close
        return tmp
      end
    end

    # Write out the barcode to a PNG image. If no filename is given, output will be returned as a string.
    def to_png fname = nil
      if fname then 
        fname += ".png" unless (fname =~ /\.png$/i )
      end
      return to_imgformat(fname, "PNG")
    end

    # Write out the barcode to a GIF image. If no filename is given, output will be returned as a string.
    def to_gif fname = nil
      if fname then 
        fname += ".gif" unless (fname =~ /\.gif$/i )
      end
      return to_imgformat( fname, "GIF")
    end

    # Write out the barcode to a JPEG image. If no filename is given, output will be returned as a string.
    def to_jpg fname = nil
      if fname then 
        fname += ".jpg" unless (fname =~ /\.jpg$/i )
      end
      return to_imgformat(fname, "JPG")
    end

    private :enc_to_const, :create_image, :to_imgformat    
  end
end